// fp-ts/lib/TaskThese.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.taskThese = exports.mapLeft = exports.map = exports.bimap = exports.toTuple = exports.getMonad = exports.getSemigroup = exports.swap = exports.fold = exports.fromIOEither = exports.rightTask = exports.leftTask = exports.leftIO = exports.rightIO = exports.both = exports.right = exports.left = exports.URI = void 0;
var Task_1 = require("./Task");
var TH = require("./These");
var TheseT_1 = require("./TheseT");
var T = /*#__PURE__*/ TheseT_1.getTheseM(Task_1.monadTask);
/**
 * @since 2.4.0
 */
exports.URI = 'TaskThese';
/**
 * @since 2.4.0
 */
exports.left = T.left;
/**
 * @since 2.4.0
 */
exports.right = T.right;
/**
 * @since 2.4.0
 */
exports.both = T.both;
/**
 * @since 2.4.0
 */
function rightIO(ma) {
    return exports.rightTask(Task_1.fromIO(ma));
}
exports.rightIO = rightIO;
/**
 * @since 2.4.0
 */
function leftIO(me) {
    return exports.leftTask(Task_1.fromIO(me));
}
exports.leftIO = leftIO;
/**
 * @since 2.4.0
 */
exports.leftTask = T.leftM;
/**
 * @since 2.4.0
 */
exports.rightTask = T.rightM;
/**
 * @since 2.4.0
 */
exports.fromIOEither = Task_1.fromIO;
/**
 * @since 2.4.0
 */
function fold(onLeft, onRight, onBoth) {
    return function (fa) { return T.fold(fa, onLeft, onRight, onBoth); };
}
exports.fold = fold;
/**
 * @since 2.4.0
 */
exports.swap = T.swap;
/**
 * @since 2.4.0
 */
function getSemigroup(SE, SA) {
    return Task_1.getSemigroup(TH.getSemigroup(SE, SA));
}
exports.getSemigroup = getSemigroup;
/**
 * @since 2.4.0
 */
function getMonad(S) {
    return __assign(__assign({ URI: exports.URI }, T.getMonad(S)), { fromIO: rightIO, fromTask: exports.rightTask });
}
exports.getMonad = getMonad;
/* tslint:disable:readonly-array */
/**
 * @since 2.4.0
 */
function toTuple(e, a) {
    return function (fa) { return T.toTuple(fa, e, a); };
}
exports.toTuple = toTuple;
/* tslint:enable:readonly-array */
// -------------------------------------------------------------------------------------
// pipeables
// -------------------------------------------------------------------------------------
/**
 * @since 2.4.0
 */
exports.bimap = function (f, g) { return function (fa) { return T.bimap(fa, f, g); }; };
/**
 * @since 2.4.0
 */
exports.map = function (f) { return function (fa) { return T.map(fa, f); }; };
/**
 * @since 2.4.0
 */
exports.mapLeft = function (f) { return function (fa) {
    return T.mapLeft(fa, f);
}; };
// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------
/**
 * @since 2.4.0
 */
exports.taskThese = {
    URI: exports.URI,
    map: T.map,
    bimap: T.bimap,
    mapLeft: T.mapLeft
};
